geo-traits/src/to_geo.rs

```diff
@@ -0,0 +1,163 @@
+//! Convert structs that implement geo-traits to [geo-types] objects.
+
+use geo_types::{
+    Coord, CoordNum, Geometry, GeometryCollection, Line, LineString, MultiLineString, MultiPoint,
+    MultiPolygon, Point, Polygon, Rect, Triangle,
+};
+
+use crate::{
+    CoordTrait, GeometryCollectionTrait, GeometryTrait, GeometryType, LineStringTrait, LineTrait,
+    MultiLineStringTrait, MultiPointTrait, MultiPolygonTrait, PointTrait, PolygonTrait, RectTrait,
+    TriangleTrait,
+};
+
+/// Convert any coordinate to a [`Coord`].
+///
+/// Only the first two dimensions will be kept.
+pub fn coord_to_geo<T: CoordNum>(coord: &impl CoordTrait<T = T>) -> Coord<T> {
```

Comment:
Since these functions are exported from the `to_geo` mod, should we name it `to_coord` instead of `coord_to_geo`?

```rs
use geo_traits::to_geo;

to_geo::to_coord(...);
to_geo::to_point(...);
```
etc.

Reply:
Or should that be: `to_geo::from_coord(some_coord_trait)` 🤯😱💥

Reply:
To be clear, I don't really know. I do think your shorter name is better than what's there current and unlikely to be confusing in context (assuming the caller keeps the `to_geo` in scope.

Reply:
Either way. I don't have a preference. Maybe `from_coord` is preferable in the hope that some day in the future the traits are stable enough that we could have `geo_types::Coord::from_coord`?

Reply:
You could do that today if you introduced a trait:

in geo-traits crate:
```
impl FromPointTrait {
    fn from_point(point: impl PointTrait) -> Self;
}

impl FromPointTrait on geo::Point {
...
 }
```

And then also in other crates:
```
impl FromPointTrait for wkt::Point {}
```

Then you could do things like:
```
let geo_point = geo::Point::new(1.0, 2.0);
let wkt_point = wkt::Point::from_point(geo_point);
let back_to_geo_point = geo::Point::from(wkt_point);
```

Is that useful? I dunno. For some formats it'd be a nice conversion muxer between formats. Though I don't know that it would work for all formats where there isn't a "standalone" structure for one geometry (e.g. converting a something to a single standalone GeoArrow point might be weird, right?)

Reply:
I see these converter functions as side-stepping the question of "how do you _construct_ an object that implements these traits". For now, these converters are a useful "bridge" until some uncertain point in the future when geo's APIs are implemented in terms of traits.

I've [argued before](https://github.com/georust/geo/pull/1157#issuecomment-1971835800) that I don't think the traits should necessarily provide a way to construct new objects. I think that should be implementation defined.

Reply:
Rather I could imagine a trait being defined like 

```rs
/// Convert to a geo_types [`Coord`].
pub trait ToGeoCoord<T: CoordNum> {
    /// Convert to a geo_types [`Coord`].
    fn to_coord(&self) -> Coord<T>;
}

impl<T: CoordNum, G: CoordTrait<T = T>> ToGeoCoord<T> for G {
    fn to_coord(&self) -> Coord<T> {
        Coord {
            x: self.x(),
            y: self.y(),
        }
    }
}
```

I.e. the point isn't to enable construction of _anything that implements these traits_. It's to enable the construction of _geo-types_ specifically.

Reply:
> Rather I could imagine a trait being defined like

I like that. I'm happy with either merging something like `ToGeoCoord` or `to_coord` as you previously phrased.

Reply:
> I've https://github.com/georust/geo/pull/1157#issuecomment-1971835800 that I don't think the traits should necessarily provide a way to construct new objects. I think that should be implementation defined.

In any case, I'm not advocating for adding a `new` method to CoordTrait, like you were arguing against in that comment, but rather to add a new trait designating "This thing can be constructed from a `impl geo-traits`" 

I think that's basically the `From` foil of what you're proposing in `To` form with `ToGeoCoord`.

Reply:
Right, that's an important distinction. Especially because the `From` could be optionally implemented and isn't a requirement to implement the trait.

> > Rather I could imagine a trait being defined like
> 
> I like that. I'm happy with either merging something like `ToGeoCoord` or `to_coord` as you previously phrased.

Happy to do whatever. Any preferences?

Reply:
Dealers choice.

Reply:
I switched to traits. Do you like this impl?

Reply:
Looks good to me!